core/src/main/scala/hedgehog/core/Tree.scala

```diff
@@ -5,8 +5,12 @@ import hedgehog.predef._
 /**
  * NOTE: This differs from the Haskell version by not having an effect on the `Node` for performance reasons.
  * See `haskell-difference.md` for more information.
+ *
+ * FIXME The `LazyList` here is critical to avoid running extra tests during shrinking.
+ * The alternative might be something like:
+ * https://github.com/hedgehogqa/scala-hedgehog/compare/topic/issue-66-lazy-shrinking
  */
-case class Tree[M[_], A](value: A, children: M[List[Tree[M, A]]]) {
+case class Tree[M[_], A](value: A, children: M[LazyList[Tree[M, A]]]) {
```

Comment:
We're actually discussing about doing the opposite in the F# version: https://github.com/hedgehogqa/fsharp-hedgehog/pull/180/files#diff-440c70d7e172f54e93d63abc97ef0a55

Definitely a good reason (for me) to look at #66 and investigate further on this.

Reply:
@moodmosaic Yeah Jacob mentioned your PR after I had merged all of mine. At the moment I'm just happy that my shrinking doesn't run the wrong tests _and_ it runs in constant memory. I suspect I've made it more complicated than I need to though so I'm curious what you find on your side of the fence. :)

Reply:
@charleso I'll keep you posted. :+1: